cors/handler.go

```diff
@@ -0,0 +1,94 @@
+package cors
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+)
+
+type preflightInfo struct {
+	allowedMethods []string
+	allowedHeaders []string
+}
+
+type Middleware struct {
+	paths map[string]preflightInfo
+}
+
+func New() *Middleware {
+	return &Middleware{
+		paths: map[string]preflightInfo{},
+	}
+}
+
+func (m *Middleware) Middleware() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Access-Control-Allow-Origin", r.Header.Get("Origin"))
+			if r.Method == http.MethodOptions {
+				path, err := mux.CurrentRoute(r).GetPathTemplate()
```

Comment:
The other middlewares matches on the route itself, which works since Gorilla sends you the same pointer, maybe we should use the same thing in all middlewares?

Reply:
Here we need to use the path; since we have one handler per method, we register separate `Route`s for each HTTP method.